src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -0,0 +1,178 @@
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' This module contains the IntrinsicCall node implementation.'''
+
+from collections import namedtuple
+from enum import Enum
+
+from psyclone.psyir.nodes.call import Call
+from psyclone.psyir.nodes.reference import Reference
+from psyclone.psyir.symbols import IntrinsicSymbol
+
+
+class IntrinsicCall(Call):
+    ''' Node representing a call to an intrinsic routine (function or
+    subroutine). This can be found as a standalone statement
+    or an expression.
+
+    '''
+    # Textual description of the node.
+    _children_valid_format = "[DataNode]*"
+    _text_name = "IntrinsicCall"
+    _colour = "cyan"
+    _min_arg_count = 0
+    _max_arg_count = 0
+
+    #: The intrinsics that can be represented by this node.
+    Intrinsic = Enum('Intrinsic', [
+        'ALLOCATE', 'DEALLOCATE', 'RANDOM'
+        ])
+    #: Named tuple for describing the properties of the required arguments to
+    #: a particular intrinsic. If there's no limit on the number of arguments
+    #: then `max_count` will be None.
+    ArgDesc = namedtuple('ArgDesc', 'min_count max_count types')
+    #: List of ArgDesc objects describing the required arguments, indexed
+    #: by intrinsic name.
+    _required_args = {}
+    #: Dict of optional arguments, indexed by intrinsic name. Each optional
+    #: argument is described by an entry in the Dict.
+    _optional_args = {}
+    # The positional arguments to allocate must all be References (or
+    # ArrayReferences but they are a subclass of Reference).
+    _required_args[Intrinsic.ALLOCATE.name] = ArgDesc(1, None, Reference)
+    _optional_args[Intrinsic.ALLOCATE.name] = {"mold": Reference,
+                                               "stat": Reference}
+    _required_args[Intrinsic.DEALLOCATE.name] = ArgDesc(1, None, Reference)
+    _optional_args[Intrinsic.DEALLOCATE.name] = {"stat": Reference}
+    _required_args[Intrinsic.RANDOM.name] = ArgDesc(1, 1, Reference)
+    _optional_args[Intrinsic.RANDOM.name] = {}
+
+    @classmethod
+    def create(cls, intrinsic, arguments):
+        '''Create an instance of class cls given the type of intrinsic and a
+        list of nodes (or name-and-node tuples) for its arguments. Any
+        named arguments *must* come after any required arguments.
+
+        :param intrinsic: the Intrinsic being called.
+        :type intrinsic: py:class:`psyclone.psyir.IntrinsicCall.Intrinsic`
+        :param arguments: the arguments to this routine, and/or \
+            2-tuples containing an argument name and the \
+            argument. Arguments are added as child nodes.
+        :type arguments: List[ \
+            Union[:py:class:``psyclone.psyir.nodes.DataNode``, \
+                  Tuple[str, :py:class:``psyclone.psyir.nodes.DataNode``]]]
```

Comment:
Any reason for the double black-quote? We usually use the single-backquote for type expressions, but I don't know if that matters.

Reply:
No idea why I did that  - gone now.